### DISCUSSION TITLE
"Is this building existing" from external database - is it a good idea?

### DISCUSSION THREAD

**Discussion by matkoniecz** [streetcomplete/StreetComplete#2477] [Q&A] [answered]
https://budynki.openstreetmap.org.pl/ is a website made by Polish community for integrating official data with OSM.

Part of the required work is verifying buildings/addresses and adding them to OSM. Data is generally very good, but some verification is needed.

Would it be OK quest to have quest asking whatever specific geometry can be added as a building?

There is no API right now, but author and maintainer of this site proposed this idea on their own and they would be happy to implement it if it would be used by SC or something else.

Legend for website: purple dots are addresses and red areas are buildings, both are filtered to not duplicate existing ones.

There is already support for "this does not exists, remove it from objects proposed for import" but there is no API (it would be created if there would be potential for that).

**Comment by Cj-Malone** [comment]
This may be one that fits better with MapRoulette. There is a new challenge type that loads an object into a editor, that then could be alligned and uploaded.

**Reply by peternewman** [reply]
Is that perhaps true for the buildings, but checking the addresses on the ground might make sense?

**Reply by matkoniecz** [reply]
Often remote editing for buildings is feasible, and bulk of edits is done - though using JOSM. Is MapRoulette even supporting geometry imports?

**Reply by Cj-Malone** [reply]
Yeah, check out [coop challenges](https://youtu.be/RhPFAtWwhkw?t=814), I think making the challenge is bit more involved than normal ones, but it is something they support.

**Comment by peternewman** [comment]
I wonder if there is some benefit to designing some sort of generic API for things like this, effectively MapRoulette on the ground?

Some of @rjw62 's OSM tools could potentially be a good fit with this too:
https://osm.mathmos.net/

A prime example being this:
https://osm.mathmos.net/ghosts/

Although I suspect ideally integrating that data into NSI or something would be best, then it could work across other countries too.

**Comment by andrewharvey** [comment]
Can you just add notes to the map? SC will then ask users to answer the note, once it's answered then a mapper can use a desktop editor to complete the change per the note response.

**Reply by matkoniecz** [reply]
Notes cannot contain geometry, adding thousands/millions of notes is bad idea, if mapped using other methods note cleanup would be obnoxious, automatically generated notes are not accepted in general, such mass addition would swamp manually created real notes.

**Reply by andrewharvey** [reply]
They don't need to, they can just be a point at the centroid. I thought this was just a handful of buildings detected as the diff to ones already mapped, but seems you're talking about a mass import. You could just go ahead and import all the buildings, then SC will ask a bunch of quests about the building, and the user could always answer building does not exist to find the few that don't.

**Reply by matkoniecz** [reply]
There are areas where mass import is the best solution, but even there mapping them in this way would be useful. People familiar  with and being able to handle JOSM in general already imported this in their area. For remaining ones import in this way is much better than not adding it at all or waiting until people map it in their area.

But I can get back once era of mass import ends and we will get to gradual updates where aerial is no longer a good indicator.

There is already a growing number of places where only some buildings are missing and often distinguishing between demolished and not mapped buildings is not possible from aerials alone, see https://budynki.openstreetmap.org.pl/#map=16.71/50.008344/20.039767

Buildings in red are ones missing in OSM, present in official data

![screen01](https://user-images.githubusercontent.com/899988/110280829-c84d3300-7fdb-11eb-9e99-f577d375d2d7.png)

**Comment by mnalis** [comment]
- For building outlines, it would seem to me that best way is to verify them in JOSM/iD against aerial / satellite (on the ground one might _sometimes_ be able to verify existence of building, but it would be much harder/impossible to verify if geometry is correct). You should probably using something like (local) [Tasking manager](https://tasks.hotosm.org/) to keep track of which area was checked, and which wasn't (yet)

- For addresses, on-the-ground survey would seem to be the best. If the tool would do redundancy checking with wider radius (for example, [21](https://budynki.openstreetmap.org.pl/#map=18.52/52.5379399/19.7646835) seems to be duplicate / same building?) Then it might be fine just to import all missing addresses, marked with `source=whatever` and `fixme=needs verification on the ground` tags? 
Then SC perhaps could be modified to look not only for OSM Notes, but for `fixme` tags too, which would help to faciliate on-the-ground checks (and would have wider usability than just for this specific case), and the users could open Notes for (hopefully just few) things which are wrong on the ground and need full blown editor to fix them.

**Comment by westnordost** [comment, marked as answer]
Hmm, I didn't answer on it though the question was probably directed at me. Generally, sourcing from other sources is always possible as long as it there will not be a problem with the license of the produced work. And then of course, as you know, any quest generated from that must suffice to the quest guidelines. As long as this is given, I see nothing that speaks against it in principle.

Of course, to implement a quest just for one country is always something where you need to consider if it is worth the effort.

Also, as a side node, it is and always has been possible to have quests that are not directly editing OSM data but speak with some other API, albeit with considerable effort. As a matter of fact, this is already the case: OSM notes are not part of the OSM data API proper, it could as well have been a different service alltogether. Yet, they are displayed as quests.

**Reply by matkoniecz** [reply]
> Of course, to implement a quest just for one country is always something where you need to consider if it is worth the effort.

I would say yes, it is relying on database with ongoing updates - and while initial mass import can be more easily done with JOSM handling new buildings with remote editing using aerial imagery is basically impossible.

And most of code should be reusable for similar quests in other places.

Though it depends on how annoying it would be to implement.

**Reply by matkoniecz** [reply]
Looking at it again: in case where there are hundreds of buildings it may be faster to import them in JOSM.

But...

Surveying in place would allow easier verification of tree-covered ones (and harder verification of some geometry issues) and even if less optimal it would still be the best available contribution in a given place.

Main risk that I see is a new building getting added on top of no longer existing footway/forest/road and old data not deleted. User may open note but some may skip it, and even with note open it may be suboptimal.

**Reply by ttomasz** [reply]
Hi @westnordost I prepared sample endpoints if you wish to test integration between StreetComplete and https://budynki.openstreetmap.org.pl

They return GeoJSON with addresses/buildings in requested area that were marked as possibly missing in OSM i.e. they are in government databases but not in OSM.

For addresses:

GET https://budynki.openstreetmap.org.pl/sc/proposed_addresses
bounding box as parameters: xmin, ymin, xmax, ymax
example: https://budynki.openstreetmap.org.pl/sc/proposed_addresses?xmin=19.7790491118258&ymin=52.51373101834943&xmax=19.785400061596533&ymax=52.516928496892206

On my site I allow users to "report" objects that are not valid or no longer existing which deletes them from the database. This endpoint would allow StreetComplete to provide feedback back to my site.
POST https://budynki.openstreetmap.org.pl/sc/proposed_addresses/report
body with a list of addresses' ids to report (delete), example: ["2e3b47ee-873c-441e-bb3e-57a2df0d1df3", "52f60add-3710-440e-a737-b2df32bfe380"]

For buildings:

GET https://budynki.openstreetmap.org.pl/sc/proposed_buildings
example: https://budynki.openstreetmap.org.pl/sc/proposed_buildings?xmin=19.7790491118258&ymin=52.51373101834943&xmax=19.785400061596533&ymax=52.516928496892206

POST https://budynki.openstreetmap.org.pl/sc/proposed_buildings/report

If different structure would be better I can adjust so please let me know.

**Reply by mnalis** [reply]
@ttomasz that sounds very interesting, but the discussion seems to have been stopped, probably forgotten due to all the other things to do! Is the source for it available? 

We're getting some [building vector datasets for Croatia](https://tms.osm-hr.org/rijeka-buildings-bw/#18/45.33515/14.42479) too, so it sounds interesting for `.osm` manual inspection / import for things which are visible and have recent aerials; but even more so  if StreetComplete (or [SCEE](https://wiki.openstreetmap.org/wiki/SCEE)) were to allow to specify the server to use for that quite-generic-looking API (or perhaps would you be open to hosting Croatian data too? and potentially others in the future?) for things which don't have recent aerial images or are hidden under canopy.

**Reply by ttomasz** [reply]
@mnalis Well, it is: https://github.com/openstreetmap-polska/gugik2osm but to be honest I would not recommend trying to use that.

The site has been working in a stable manner all this time and recently has been heavily used to automatically verify existence of buildings on orthophotoimagery (which might be another project that will interest you): https://www.openstreetmap.org/user/NorthCrab/diary/401999 but the code for the site itself is quite messy, I never finished migrating the orchestration of processes to airflow and the database follows the specific schema of Polish data so it would be difficult to adjust it to new datasets.

As far as I can tell StreetComplete is not interested in implementing support for that API.

I would recommend writing something new that will work with your data. At first I started with just generating some files (I'd recommend with just writing GeoJSON files since JOSM can read them).

I can try to provide some guidance over tools available etc. Also if you prepare the data according to guidance by Esri you may be able to get it included in RapiD datasets which would also make it easier for people who don't use JOSM to contribute.

If you use Discord you can find me at openstreetmap.org.pl/discord (nick: tomczk, "importy" channel) otherwise you can make an issue in the repo that I mentioned or write an e-mail to tomasz.taras at openstreetmap.org.pl.

**Reply by matkoniecz** [reply]
> As far as I can tell StreetComplete is not interested in implementing support for that API.

I am interested but it is a big enough project that it is tricky to do it. Right now I am waiting for grant request for funding development of related but simpler feature that could be stepping stone for that.

**Reply by Helium314** [reply]
I tried implementing the Osmose quest in SCEE in a generic way, so other quests based on external APIs can be added with reduced effort.

**Reply by mnalis** [reply]
Thanks everyone for valuable feedback, it will likely be some time before Croatia community starts to look deeper into it (we have [just got the vectors](https://community.openstreetmap.org/t/osm-meetup-2023-11-09-online/105853) and there are other things with higher priority to do. But there is certainly the interest in finding the best (and easiest) solution to use it!

When new things are implemented or get moving, I'd much appreciate if you'd post / link to this discussion again!
